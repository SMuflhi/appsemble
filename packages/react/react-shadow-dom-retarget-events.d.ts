declare module 'react-shadow-dom-retarget-events' {
  export default function retargetEvents(shadowRoot: ShadowRoot): void;
}
