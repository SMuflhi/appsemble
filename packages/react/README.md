# Appsemble React SDK

> Build your own blocks using React

## Installation

```sh
npm install @appsemble/react
```

## Hello world example

<!-- eslint-disable import/no-extraneous-dependencies, react/jsx-filename-extension -->

```js
import { bootstrap } from '@appsemble/react';
import PropTypes from 'prop-types';
import React from 'react';


class MyBlock extends React.Component {
  static propTypes = {
    actions: PropTypes.shape().isRequired,
  };

  render() {
    const {
      actions,
    } = this.props;

    return (
      <button
        type="button"
        onClick={actions.click.dispatch}
      >
        Hello world!
      </button>
    );
  }
}


bootstrap(MyBlock);
```

**Note**: The script will be loaded only once. The component is then bootstrapped for every instance of your block type that is loaded by the app creator.
