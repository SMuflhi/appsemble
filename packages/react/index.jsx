import { bootstrap as sdkBootstrap } from '@appsemble/sdk';
import React from 'react';
import { render } from 'react-dom';
import retargetEvents from 'react-shadow-dom-retarget-events';

const { Consumer, Provider } = React.createContext();

/**
 * Mount a React component returned by a bootstrap function in the shadow DOM of a block.
 */
export function mount(Component, root) {
  return params => {
    const reactRoot = params.shadowRoot.appendChild(
      root ? root.cloneNode() : document.createElement('div'),
    );
    const props = {
      ...params,
      reactRoot,
    };
    const component = (
      <Provider value={props}>
        <Component {...props} />
      </Provider>
    );
    render(component, reactRoot);
    /**
     * React doesnâ€™t play nice with shadow DOM. This library works around that. However, the
     * implementation does contain some bugs.
     *
     * Related issues:
     *
     * - https://github.com/facebook/react/issues/2043
     * - https://github.com/facebook/react/issues/9242
     * - https://github.com/facebook/react/issues/10422
     * - https://github.com/facebook/react/issues/12973
     * - https://github.com/facebook/react/pull/8117
     * - https://github.com/facebook/react/pull/12163
     *
     * @todo Remove this when React has fixes this issue.
     */
    retargetEvents(params.shadowRoot);
  };
}

export function bootstrap(Component, reactRoot) {
  return sdkBootstrap(mount(Component, reactRoot));
}

/**
 * A HOC which passes the Appsemble block values to he wrapped React component.
 */
export function withBlock(Component) {
  function Wrapper(props) {
    return <Consumer>{values => <Component {...values} {...props} />}</Consumer>;
  }
  if (process.env.NODE_ENV !== 'production') {
    Wrapper.displayName = `withBlock(${Component.displayName || Component.name})`;
  }
  return Wrapper;
}
