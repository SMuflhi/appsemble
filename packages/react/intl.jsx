import React from 'react';
import {
  IntlProvider,
} from 'react-intl';


/**
 * A HOC which provides `intl` in a block written in React.
 */
// eslint-disable-next-line import/prefer-default-export
export function provideIntl(Component) {
  return class extends React.Component {
    render() {
      return (
        <IntlProvider locale="en-US" defaultLocale="en-US" textComponent={React.Fragment}>
          <Component {...this.props} />
        </IntlProvider>
      );
    }
  };
}
