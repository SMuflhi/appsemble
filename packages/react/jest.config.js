module.exports = {
  displayName: '@appsemble/react',
  moduleNameMapper: {
    [/@appsemble\/([\w-]+)/.source]: '@appsemble/$1/src',
  },
  setupFilesAfterEnv: ['<rootDir>/jest.setup.ts'],
  snapshotSerializers: ['enzyme-to-json/serializer'],
};
