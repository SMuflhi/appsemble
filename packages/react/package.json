{
  "name": "@appsemble/react",
  "version": "0.9.1",
  "dependencies": {
    "react-shadow-dom-retarget-events": "^1.0.10"
  },
  "peerDependencies": {
    "@appsemble/sdk": "^0.9.1",
    "react": "^16.11.0",
    "react-dom": "^16.11.0",
    "react-intl": "^3.4.0"
  }
}
