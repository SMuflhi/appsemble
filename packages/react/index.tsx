import { bootstrap as sdkBootstrap, BootstrapParams } from '@appsemble/sdk';
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import retargetEvents from 'react-shadow-dom-retarget-events';

export interface BlockProps extends BootstrapParams {
  /**
   * The DOM node on which the block is mounted.
   */
  reactRoot: HTMLElement;
}

const { Consumer, Provider } = React.createContext<BlockProps>(null);

/**
 * Mount a React component returned by a bootstrap function in the shadow DOM of a block.
 */
export function mount(
  Component: React.ComponentType<BlockProps>,
  root?: HTMLElement,
): (params: BootstrapParams) => void {
  return params => {
    const reactRoot = params.shadowRoot.appendChild(
      root ? root.cloneNode() : document.createElement('div'),
    ) as HTMLElement;
    const props = {
      ...params,
      reactRoot,
    };
    const component: JSX.Element = (
      <Provider value={props}>
        <Component {...props} />
      </Provider>
    );
    ReactDOM.render(component, reactRoot);
    params.utils.addCleanup(() => ReactDOM.unmountComponentAtNode(reactRoot));
    /**
     * React doesnâ€™t play nice with shadow DOM. This library works around that. However, the
     * implementation does contain some bugs.
     *
     * Related issues:
     *
     * - https://github.com/facebook/react/issues/2043
     * - https://github.com/facebook/react/issues/9242
     * - https://github.com/facebook/react/issues/10422
     * - https://github.com/facebook/react/issues/12973
     * - https://github.com/facebook/react/pull/8117
     * - https://github.com/facebook/react/pull/12163
     *
     * @todo Remove this when React has fixes this issue.
     */
    retargetEvents(params.shadowRoot);
  };
}

export function bootstrap(
  Component: React.ComponentType<BlockProps>,
  reactRoot?: HTMLElement,
): void {
  sdkBootstrap(mount(Component, reactRoot));
}

/**
 * A HOC which passes the Appsemble block values to he wrapped React component.
 */
export function withBlock<P extends object>(
  Component: React.ComponentType<P & Omit<BlockProps, keyof P>>,
): React.ComponentType<P> {
  function Wrapper(props: P): JSX.Element {
    return <Consumer>{values => <Component {...values} {...props} />}</Consumer>;
  }
  if (process.env.NODE_ENV !== 'production') {
    Wrapper.displayName = `withBlock(${Component.displayName || Component.name})`;
  }
  return Wrapper;
}
