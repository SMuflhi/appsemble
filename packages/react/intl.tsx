import * as React from 'react';
import { IntlProvider } from 'react-intl';

/**
 * A HOC which provides `intl` in a block written in React.
 */
// eslint-disable-next-line import/prefer-default-export
export function provideIntl(
  Component: React.ComponentType,
  intlProviderProps: IntlProvider.Props,
): React.ComponentType {
  return props => (
    <IntlProvider
      defaultLocale="en-US"
      locale="en-US"
      {...intlProviderProps}
      textComponent={React.Fragment}
    >
      <Component {...props} />
    </IntlProvider>
  );
}
