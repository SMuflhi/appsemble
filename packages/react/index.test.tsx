import { BootstrapParams } from '@appsemble/sdk';
import * as enzyme from 'enzyme';
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import retargetEvents from 'react-shadow-dom-retarget-events';

import { mount, withBlock } from '.';

jest.mock('react-dom', () => ({
  render: jest.fn(),
}));

jest.mock('react-shadow-dom-retarget-events', () => jest.fn());

let bootstrapParams: BootstrapParams;

beforeEach(() => {
  bootstrapParams = {
    actions: {},
    block: {
      type: 'test',
      version: '0.0.0',
    },
    data: {},
    events: {
      emit() {},
      off() {},
      on() {},
    },
    shadowRoot: {
      appendChild(node: HTMLElement) {
        return node;
      },
    } as any,
    utils: {
      showMessage() {},
    },
  };
});

describe('mount', () => {
  it('should mount the react component', () => {
    let component: JSX.Element;
    (ReactDOM.render as jest.Mock).mockImplementation(c => {
      component = c;
    });
    mount(() => <div>test</div>, document.createElement('div'))(bootstrapParams);
    expect(enzyme.mount(component)).toMatchSnapshot();
  });

  it('should retarget DOM events to the shadow root', () => {
    mount(() => <div>test</div>, document.createElement('div'))(bootstrapParams);
    expect(retargetEvents).toHaveBeenCalledWith(bootstrapParams.shadowRoot);
  });
});

describe('withBlock', () => {
  it('should pass block props to the child element', () => {
    const Test = withBlock(function Test() {
      return null;
    });
    let component: JSX.Element;
    (ReactDOM.render as jest.Mock).mockImplementation(c => {
      component = c;
    });
    mount(() => <Test />, document.createElement('div'))(bootstrapParams);
    expect(enzyme.mount(component)).toMatchSnapshot();
  });
});
