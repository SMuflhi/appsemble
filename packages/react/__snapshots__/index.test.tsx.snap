// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mount should mount the react component 1`] = `
<Component
  actions={Object {}}
  block={
    Object {
      "type": "test",
      "version": "0.0.0",
    }
  }
  data={Object {}}
  events={
    Object {
      "emit": [Function],
      "off": [Function],
      "on": [Function],
    }
  }
  reactRoot={<div />}
  shadowRoot={
    Object {
      "appendChild": [Function],
    }
  }
  utils={
    Object {
      "showMessage": [Function],
    }
  }
>
  <div>
    test
  </div>
</Component>
`;

exports[`withBlock should pass block props to the child element 1`] = `
<Component
  actions={Object {}}
  block={
    Object {
      "type": "test",
      "version": "0.0.0",
    }
  }
  data={Object {}}
  events={
    Object {
      "emit": [Function],
      "off": [Function],
      "on": [Function],
    }
  }
  reactRoot={<div />}
  shadowRoot={
    Object {
      "appendChild": [Function],
    }
  }
  utils={
    Object {
      "showMessage": [Function],
    }
  }
>
  <withBlock(Test)>
    <Test
      actions={Object {}}
      block={
        Object {
          "type": "test",
          "version": "0.0.0",
        }
      }
      data={Object {}}
      events={
        Object {
          "emit": [Function],
          "off": [Function],
          "on": [Function],
        }
      }
      reactRoot={<div />}
      shadowRoot={
        Object {
          "appendChild": [Function],
        }
      }
      utils={
        Object {
          "showMessage": [Function],
        }
      }
    />
  </withBlock(Test)>
</Component>
`;
